react/hooks/useConfirmExit.js

```diff
@@ -0,0 +1,30 @@
+import { useCallback, useRef } from 'react'
+import useEventListener from 'cozy-ui/transpiled/react/hooks/useEventListener'
+
+/**
+ * When provided a message, will warn the user before exiting the page
+ *
+ * Warning: The confirmation message is required but will usually
+ * be replaced by an internal message by the browser, and never displayed.
+ *
+ * A falsy error message would deactivate to confirmation popup.
+ *
+ * @param {string|null} message - Confirmation message
+ * @param {function} callback - will be executed before returning
+ */
+export default function useConfirmExit(message, callback) {
+  // Using a ref in order to have an event listener that does not
+  // need to be deregistered, recreated and registered again at each
+  // message or callback change. If not, the lag introduced by the
```

Comment:
I don't really understand why we we have use a ref here ? How does it work ?

Reply:
Without the ref, the `beforeunload` event handler would change each time you change the message or the callback. Then each time it changes, it would call the `useEventListener` that would deregister the previous event handler and register the new one.

In cozy-notes, we may change the parameters very often (on a box with low latency, it may be every 20ms). It would trigger useless attachEvent and removeEvent that many times.

Furthermore, useEffet inside useEventListener is not immediate. It should be very quick but when the state change so often, I had occurence where the confirmation did not occur  because the corresponding effect still hasn't run (and an old eventhandler with obsolete parameters was still attached).

Using a stable event handler and a reference allows us to always have an handler that will use the most recent values

Reply:
After reading in more detail the useRef doc, it became clear.